components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt

```diff
@@ -280,3 +330,35 @@ internal fun Pointer.getAndConsumeRustString(): String {
 internal fun Pointer.getRustString(): String {
     return this.getString(0, "utf8")
 }
+
+/**
+ * A helper class for gathering basic performance metrics on different kinds of LoginsStore operation.
+ *
+ * For each type of operation, we want to measure:
+ *    - total count of operations performed
+ *    - count of operations that produced an unexpected error
+ *    - time taken to perform the operation
+ *
+ * This is a convenince wrapper to measure all those things in one shot.
+ */
+class LoginsStoreMetricsGroup(
+    val timer: TimingDistributionMetricType,
+    val count: CounterMetricType,
+    val errCount: CounterMetricType
+) {
+    @Suppress("TooGenericExceptionCaught")
+    inline fun <U> measure(callback: () -> U): U {
+        count.add()
+        try {
+            val timerId = timer.start()
+            try {
+                return callback()
+            } finally {
+                timer.stopAndAccumulate(timerId)
+            }
+        } catch (e: Exception) {
```

Comment:
Is it worth catching Throwable instead? Not really sure the difference, to be honest, but I believe this could fail to catch some errors (Which I'm unsure can ever exist in practice)

Reply:
Actually I wonder if I should catch something much more specific here. We probably don't want to increment this counter for "expected errors" such as `InvalidRecordException`, but only for things that went unexpectedly wrong.

Reply:
Good point. When @lougeniaC64 and I went over the design of the validation functions, I was pretty hesitant about adding a function that either 'does nothing or throws' I couldn't concretely place why, but this is part of it. Having throwing be an expected part of the use-case of the function complicates measurement (and usage, IMO).

I think at this point it might be fine to just check for InvalidRecordException and ignore it, but yeah, it's worth considering.

Reply:
I've changed this to categorize the different kinds of error via  labeled counter.